src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -0,0 +1,824 @@
+    };
+
+public:
+    using const_iterator
+      = segment_meta_frame_const_iterator<value_t, decoder_t>;
+
+    using delta_alg = delta_t;
+
+    using hint_t = deltafor_stream_pos_t<value_t>;
+
+    explicit segment_meta_column_frame(delta_alg d)
+      : _delta_alg(std::move(d)) {}
+
+    void append(value_t value) {
+        auto ix = index_mask & _size++;
+        _head.at(ix) = value;
+        if ((_size & index_mask) == 0) {
+            if (!_tail.has_value()) {
+                _tail.emplace(_head.at(0), _delta_alg);
+            }
+            _last_row = _tail->get_position();
+            _tail->add(_head);
+        }
+    }
+
+    struct frame_tx_t {
+        typename encoder_t::tx_state inner;
+        segment_meta_column_frame<value_t, delta_t>& self;
+
+        frame_tx_t(
+          typename encoder_t::tx_state&& s,
+          segment_meta_column_frame<value_t, delta_t>& self)
+          : inner(std::move(s))
+          , self(self) {}
+
+        void add(const std::array<value_t, buffer_depth>& row) {
+            inner.add(row);
+        }
+        void commit() && noexcept { self._tail->tx_commit(std::move(inner)); }
+    };
+
+    // Transactional append operation
+    //
+    // The method returns an optional with the state of the
+    // transaction. If the append operation doesn't require
+    // memory allocation the optional is null. Otherwise it
+    // will contain a transaction which need to be commited.
+    // The actual append operation can allocate memory and
+    // can throw but 'commit' method of the transaction
+    // doesn't allocate and doesn't throw.
+    std::optional<frame_tx_t> append_tx(value_t value) {
+        std::optional<frame_tx_t> tx;
+        auto ix = index_mask & _size++;
+        _head.at(ix) = value;
+        if ((_size & index_mask) == 0) {
+            if (!_tail.has_value()) {
+                _tail.emplace(_head.at(0), _delta_alg);
+            }
+            _last_row = _tail->get_position();
+            tx.emplace(_tail->tx_start(), *this);
+            tx->add(_head);
+        }
+        return tx;
+    }
+
+    /// Get stream position object which can be used as a "hint" to speed up
+    /// index lookup operations.
+    std::optional<hint_t> get_current_stream_pos() const { return _last_row; }
+
+    const_iterator at_index(size_t index) const {
+        if (index >= _size) {
+            return end();
+        }
+        auto it = begin();
+        std::advance(it, index);
+        return it;
+    }
+
+    /// Get element by index, use 'hint' to speedup the operation.
+    ///
+    /// The 'hint' has to correspond to any index lower or equal to
+    /// 'index'. The 'hint' values has to be stored externally.
+    const_iterator at_index(size_t index, const hint_t& hint) const {
+        if (index >= _size) {
+            return end();
+        }
+        decoder_t decoder(
+          _tail->get_initial_value(),
+          _tail->get_row_count(),
+          _tail->share(),
```

Comment:
potentially unsafe?

Reply:
decoder never modifies the shared buffer

Reply:
I mean accessing the optional without checking it

Reply:
It's safe because few lines above were checking that the index is not greater than the size. The `_tail` can only be `std::nullopt` if it's empty. In this case the branch will be taken and the method will return `end()`.